R/verb-fill.R

```diff
@@ -0,0 +1,195 @@
+# ignore nulls
+# * hive: https://cwiki.apache.org/confluence/display/Hive/LanguageManual+WindowingAndAnalytics
+# * impala: https://docs.cloudera.com/documentation/enterprise/5-11-x/topics/impala_analytic_functions.html
+# * mssql: https://docs.microsoft.com/en-us/sql/t-sql/functions/first-value-transact-sql?view=sql-server-ver15
+# * oracle: https://oracle-base.com/articles/misc/first-value-and-last-value-analytic-functions
+# * redshift: https://docs.aws.amazon.com/redshift/latest/dg/r_WF_first_value.html
+# * teradata: https://docs.teradata.com/r/756LNiPSFdY~4JcCCcR5Cw/V~t1FC7orR6KCff~6EUeDQ
+#
+# no ignore nulls
+# * access: https://www.techonthenet.com/access/functions/
+# * hana: https://help.sap.com/viewer/4fe29514fd584807ac9f2a04f6754767/2.0.04/en-US/e7ef7cc478f14a408e1af27fc1a78624.html
+# * mysql: https://dev.mysql.com/doc/refman/8.0/en/window-function-descriptions.html
+
+#' Fill in missing values with previous or next value
+#'
+#' @inheritParams arrange.tbl_lazy
+#' @param ... Columns to fill.
+#' @param .direction Direction in which to fill missing values. Currently
+#'   either "down" (the default) or "up".
+#'
+#' @examples
+#' squirrels <- tibble::tribble(
+#'   ~group,    ~name,     ~role,     ~n_squirrels, ~ n_squirrels2,
+#'   1,      "Sam",    "Observer",   NA,                 1,
+#'   1,     "Mara", "Scorekeeper",    8,                NA,
+#'   1,    "Jesse",    "Observer",   NA,                NA,
+#'   1,      "Tom",    "Observer",   NA,                 4,
+#'   2,     "Mike",    "Observer",   NA,                NA,
+#'   2,  "Rachael",    "Observer",   NA,                 6,
+#'   2,  "Sydekea", "Scorekeeper",   14,                NA,
+#'   2, "Gabriela",    "Observer",   NA,                NA,
+#'   3,  "Derrick",    "Observer",   NA,                NA,
+#'   3,     "Kara", "Scorekeeper",    9,                 10,
+#'   3,    "Emily",    "Observer",   NA,                NA,
+#'   3, "Danielle",    "Observer",   NA,                NA
+#' )
+#' squirrels$id <- 1:12
+#'
+#' if (require("tidyr", quietly = TRUE)) {
+#'   tbl_memdb(squirrels) %>%
+#'     window_order(id) %>%
+#'     tidyr::fill(
+#'       n_squirrels,
+#'       n_squirrels2,
+#'     )
+#' }
+fill.tbl_lazy <- function(.data, ..., .direction = c("down", "up")) {
+  sim_data <- simulate_vars(.data)
+  cols_to_fill <- syms(names(tidyselect::eval_select(expr(c(...)), sim_data)))
+  order_by_cols <- op_sort(.data)
+  .direction <- arg_match0(.direction, c("down", "up"))
+
+  if (is_empty(order_by_cols)) {
+    abort(
+      c(
+        x = "`.data` does not have explicit order.",
+        i = "Please use arrange() or window_order() to make determinstic."
+      )
```

Comment:
Can you please test this using a combination of `expect_error()` and `expect_snapshot()`? The pattern we're currently using looks something like this:

```R
expect_snapshot({
  (expect_error(df %>% fill(x))
})`

`expect_error()` ensures that an error is actually thrown, and `expect_snapshot()` verifies that it's the correct error, and gives a place to proof read the error as it will appear to the user.